Bluetooth: Mesh: Fix heartbeat sending on Friendship established/lost

jhedberg opened the pull request:
Mesh Profile 3.6.7.2 Publishing Heartbeat messages:
    
```
Triggered publishing of Heartbeat messages is enabled by
the Heartbeat Publication Features state (see Section 4.2.17.5):
...
- If the Low Power bit is set to 1, a Heartbeat message shall be
published when the node establishes or loses Friendship (see Section 3.6.6.1).
```
  
Fixes #18194

Commit:
Bluetooth: Mesh: Move heartbeat sending to transport layer

The heartbeat is a transport layer feature, so move it to transport.c.
This also opens the way to properly fix Friendship-established/lost
triggering for LPN role.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Vudentz reviewed (approved):
LGTM

carlescufi commented:
@jhedberg can you please add this to the PR with Host Conformance issues manually instead of using the backport bot?

https://github.com/zephyrproject-rtos/zephyr/pull/18122

jhedberg commented:
> @jhedberg can you please add this to the PR with Host Conformance issues manually instead of using the backport bot?

Sure, the bot will likely fail anyway since master has moved mesh from `subsys/bluetooth/host/mesh` to `subsys/bluetooth/mesh`

jhedberg commented:
@MariuszSkamra it'd be good if you could test and verify that this fixes the failing tests, the patches are now also included in #18122 which may be the better PR to run the tests on since that matches what'll end up in LTS.

trond-snekvik commented on subsys/bluetooth/mesh/lpn.c:
```diff
@@ -222,8 +222,12 @@ static void clear_friendship(bool force, bool disable)
-	if (lpn_cb && lpn->frnd != BT_MESH_ADDR_UNASSIGNED) {
-		lpn_cb(lpn->frnd, false);
+	if (lpn->frnd != BT_MESH_ADDR_UNASSIGNED) {
+		bt_mesh_heartbeat_send();
```
If I'm reading this correctly, I think there's a problem here. As `lpn_set_state()` isn't called yet, we'll be populating the heartbeat with the old LPN state here, mistakenly sending out a heartbeat with the pre-clear LPN flag, instead of the new one.

trond-snekvik commented on subsys/bluetooth/mesh/lpn.c:
```diff
@@ -958,6 +962,8 @@ int bt_mesh_lpn_friend_update(struct bt_mesh_net_rx *rx,
 		BT_INFO("Friendship established with 0x%04x", lpn->frnd);
 
+		bt_mesh_heartbeat_send();
+
```
Looks like the same issue is here, with the `friend_response_received` being called below.

jhedberg commented:
@trond-snekvik looks like you're right, good catch - I'll try to fix it

jhedberg commented:
@trond-snekvik actually you discovered a bigger issue: the code shouldn't be setting the flag based on `BT_MESH_LPN_DISABLED`. That state just tells us we're not actively trying to look for Friends, whereas the spec associates the flag with a Friendship being established or not.

jhedberg commented:
@trond-snekvik I've pushed updated (however only compile-tested) versions of the patches. Looks more reasonable?

trond-snekvik commented:
Yep, looks resolved. About this `established` flag though, isn't that just a duplication of `lpn->state >= BT_MESH_LPN_ESTABLISHED`, or is there some more nuance to it?

MariuszSkamra reviewed (changes requested):
Just as you wrote, the Heartbeat Publication is triggered by Heartbeat Publication Features state. So before you send Heartbeat message, you should check whether, 3rd bit (Low Power) is set in Heartbeat Publication Features.
```
diff --git a/subsys/bluetooth/host/mesh/lpn.c b/subsys/bluetooth/host/mesh/lpn.c
index 5ac3ad1ef3..b45b956b1e 100644
--- a/subsys/bluetooth/host/mesh/lpn.c
+++ b/subsys/bluetooth/host/mesh/lpn.c
@@ -200,6 +200,8 @@ static int send_friend_clear(void)
 static void clear_friendship(bool force, bool disable)
 {
        struct bt_mesh_lpn *lpn = &bt_mesh.lpn;
+       struct bt_mesh_cfg_srv *cfg = bt_mesh_cfg_get();
+       struct bt_mesh_subnet *sub;
 
        BT_DBG("force %u disable %u", force, disable);
 
@@ -246,7 +248,10 @@ static void clear_friendship(bool force, bool disable)
         */
        lpn->groups_changed = 1U;
 
-       bt_mesh_heartbeat_send();
+       sub = bt_mesh_subnet_get(cfg->hb_pub.net_idx);
+       if ((cfg->hb_pub.feat & BT_MESH_FEAT_LOW_POWER) && sub) {
+               bt_mesh_heartbeat_send();
+       }
 
        if (disable) {
                lpn_set_state(BT_MESH_LPN_DISABLED);
@@ -921,6 +926,7 @@ int bt_mesh_lpn_friend_update(struct bt_mesh_net_rx *rx,
 {
        struct bt_mesh_ctl_friend_update *msg = (void *)buf->data;
        struct bt_mesh_lpn *lpn = &bt_mesh.lpn;
+       struct bt_mesh_cfg_srv *cfg = bt_mesh_cfg_get();
        struct bt_mesh_subnet *sub = rx->sub;
        u32_t iv_index;
 
@@ -960,7 +966,10 @@ int bt_mesh_lpn_friend_update(struct bt_mesh_net_rx *rx,
 
                BT_INFO("Friendship established with 0x%04x", lpn->frnd);
 
-               bt_mesh_heartbeat_send();
+               if (cfg->hb_pub.feat & BT_MESH_FEAT_LOW_POWER) {
+                       bt_mesh_heartbeat_send();
+               }
 
                if (lpn_cb) {
                        lpn_cb(lpn->frnd, true);
```

with this,  MESH/NODE/CFG/HBP/BV-05-C is passing

MariuszSkamra commented on subsys/bluetooth/mesh/transport.c:
```diff
@@ -1506,3 +1506,52 @@ void bt_mesh_rpl_clear(void)
+
+	hb.init_ttl = cfg->hb_pub.ttl;
+
+	if (bt_mesh_relay_get() == BT_MESH_RELAY_ENABLED) {
```
Is that correct? Shouldn't the current cfg->hb_pub.feat status be returned? Does not mean that any test case is failing because of this but I'm just asking.

As written in 4.3.2.63 Config Hearbeat Publication Status:
"The Features field shall identify the Heartbeat Publication Features state (see Section 4.2.17.5)."

jhedberg replied:
@MariuszSkamra I always understood cfg->hb_pub.feat as the triggers for sending the message, but not for controlling the content of the message. I'll need to double-check the spec and test spec.

jhedberg replied:
4.2.17.5 Heartbeat Publication Features: "The Heartbeat Publication Features state determines the features that trigger sending Heartbeat messages when changed."

4.3.2.63 Config Heartbeat Publication Status: "Bit field indicating features that trigger Heartbeat messages when changed"

So from that perspective this code looks correct.

jhedberg commented:
> Just as you wrote, the Heartbeat Publication is triggered by Heartbeat Publication Features state. So before you send Heartbeat message, you should check whether, 3rd bit (Low Power) is set in Heartbeat Publication Features.

@MariuszSkamra good point, and thanks for the refresher on how heartbeat publication is supposed to work 😄 

I'll update the PR in a moment.

jhedberg commented:
```
+       sub = bt_mesh_subnet_get(cfg->hb_pub.net_idx);
+       if ((cfg->hb_pub.feat & BT_MESH_FEAT_LOW_POWER) && sub) {
+               bt_mesh_heartbeat_send();
```

@MariuszSkamra the checking for the subnet is redundant. The configuration server will not allow enabling heartbeat publication on a subnet we're not on, and the `bt_mesh_subnet_del()` function will implicitly disable heartbeat publication if the subnet matches. I can see you just copied the subnet check from `cfg_srv.c` but it looks like it's unnecessary there as well.

jhedberg commented:
@MariuszSkamra actually, ignore my previous comment - the check for subnet is essentially a check for whether publication is enabled or not. The subnet index will be invalid if publication is disabled. Let me see if there'd be a more intuitive way to do this check to avoid this kind of confusion of its purpose.

Commit:
Bluetooth: Mesh: Fix checking for active heartbeat publication

When sending heartbeat messages triggered by feature changes the code
was trying to look up the configured publication subnet, in an
apparent attempt to figure out if publication is enabled or not. A
more appropriate way is to check for the heartbeat publication
destination address, and since we have a helper function this can be
done in a single place.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Mesh: Fix heartbeat sending on Friendship established/lost

Mesh Profile 3.6.7.2 Publishing Heartbeat messages:

"
Triggered publishing of Heartbeat messages is enabled by
the Heartbeat Publication Features state (see Section 4.2.17.5):
...
- If the Low Power bit is set to 1, a Heartbeat message shall be
 published when the node establishes or loses Friendship (see Section
 3.6.6.1).
"

Fixes #18194

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
@MariuszSkamra I've pushed updated patches. Theres one new one for fixing the checking of active heartbeat publication. Could you take a look?

jhedberg commented:
@jhedberg wrote:
> The subnet index will be invalid if publication is disabled

This wasn't even true: the foundation model spec has no special requirements on touching the net key index value when disabling publication (I suppose since there's no "invalid" value in the spec for a key index, although we do have an internal one due to using 16 bits instead of 12). The spec only says the other values should be set to 0 when the destination is set to the unassigned address, which is why our implementation doesn't touch the key index at all in this case. So the correct way to check for enabled publication is indeed to check for the destination address.

jhedberg commented:
> About this established flag though, isn't that just a duplication of lpn->state >= BT_MESH_LPN_ESTABLISHED, or is there some more nuance to it?

@trond-snekvik no, I think you're right. That's a work for another PR however. I wouldn't do the `>= LPN_ESTABLISHED` though, but instead explicitly check for the three "established" states - that way we don't create an implicit dependency on the ordering of the enum definition.

jhedberg commented:
@trond-snekvik I take back what I just said: some of those three states are used prior to establishing friendship, so the separate established flag is needed to know what the next transition should be. E.g. we'll be in WAIT_UPDATE also for the first friend update, but the friendship isn't considered established until we get that update.

trond-snekvik commented:
I see, thanks for the clarification!
